hudi-common/src/test/java/org/apache/hudi/common/table/read/TestEventTimeMerging.java

```diff
@@ -0,0 +1,124 @@
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.hudi.common.testutils.HoodieTestDataGenerator.AVRO_SCHEMA;
+import static org.apache.hudi.common.testutils.reader.DataGenerationPlan.OperationType.DELETE;
+import static org.apache.hudi.common.testutils.reader.DataGenerationPlan.OperationType.INSERT;
+import static org.apache.hudi.common.testutils.reader.DataGenerationPlan.OperationType.UPDATE;
+import static org.apache.hudi.common.testutils.reader.HoodieFileSliceTestUtils.ROW_KEY;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TestEventTimeMerging extends HoodieFileGroupReaderTestHarness {
+  @BeforeAll
+  public static void setUp() throws IOException {
+    // Specify the key column values for each file.
+    keyRanges = Arrays.asList(
+        new HoodieFileSliceTestUtils.KeyRange(1, 10),
+        new HoodieFileSliceTestUtils.KeyRange(1, 5),
+        new HoodieFileSliceTestUtils.KeyRange(1, 3),
+        new HoodieFileSliceTestUtils.KeyRange(6, 8),
+        new HoodieFileSliceTestUtils.KeyRange(1, 2));
+    // Specify the value of `timestamp` column for each file.
+    timestamps = Arrays.asList(
+        2L, 3L, 1L, 1L, 4L);
+    // Specify the operation type for each file.
+    operationTypes = Arrays.asList(
+        INSERT, DELETE, UPDATE, DELETE, UPDATE);
+    // Specify the instant time for each file.
+    instantTimes = Arrays.asList(
+        "001", "002", "003", "004", "005");
+  }
+
+  @BeforeEach
+  public void initialize() throws Exception {
+    setTableName(TestEventTimeMerging.class.getName());
+    initPath(tableName);
+    initMetaClient();
+    initTestDataGenerator(new String[]{PARTITION_PATH});
+    testTable = HoodieTestTable.of(metaClient);
+    setUpMockCommits();
+  }
+
+  @Test
+  public void testWithOneLogFile() throws IOException, InterruptedException {
+    // The FileSlice contains a base file and a log file.
+    ClosableIterator<IndexedRecord> iterator = getFileGroupIterator(2);
+    List<String> leftKeysExpected = Arrays.asList("6", "7", "8", "9", "10");
+    List<String> leftKeysActual = new ArrayList<>();
+    while (iterator.hasNext()) {
+      leftKeysActual.add(iterator.next().get(AVRO_SCHEMA.getField(ROW_KEY).pos()).toString());
+    }
+    assertEquals(leftKeysExpected, leftKeysActual);
+  }
+
+  @Test
+  public void testWithTwoLogFiles() throws IOException, InterruptedException {
+    // The FileSlice contains a base file and two log files.
+    ClosableIterator<IndexedRecord> iterator = getFileGroupIterator(3);
+    List<String> leftKeysExpected = Arrays.asList("6", "7", "8", "9", "10");
+    List<String> leftKeysActual = new ArrayList<>();
+    while (iterator.hasNext()) {
+      leftKeysActual.add(iterator.next().get(AVRO_SCHEMA.getField(ROW_KEY).pos()).toString());
+    }
+    assertEquals(leftKeysExpected, leftKeysActual);
```

Comment:
Also validate non-key values to make sure the values are updated or kept the same based on the ordering field values?

Reply:
We can try to compare some of these non-key columns.